### PR TITLE
Basic recommender

### PR DESCRIPTION
**Pull request by junojense**
Added a framework for profiles, a coin toss recommender and a sample algorithm using the ratio between video duration and watchtime. We decided to go for aggregate watch time so the duration is currently not implemented and set to 0. Things to do on the short term are;

    sharing profiles between peers (gossiping)
    implement an algorithm for sorting the list of profiles based on yield prediction
    create some kind of framework for keeping track of watched/unwatched videos
    there are also TODOs in the code but I am not sure if we should be implementing those
    On the long term we should think about what fields we can add to the profiles (e.g. engagement, followers of person posting), how we can detect when to stop investing in a video and how to add some randomness to the video selection (probability-based selection of unwatched videos)

### CODE REVIEW

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        URIs: List<String>,
+        profiles: List<Profile>
+    ): String {
+        var (candidate, averagePercentageWatched) = Pair("", 0.0f)
+        for (URI: String in URIs) {
+            var (hits, percentageWatchedTotal) = Pair(0.0f, 0.0f)
+            for (profile: Profile in profiles) {
+                if (profile.magnets.containsKey(URI)) {
+                    val entry = profile.magnets[URI]
+                    percentageWatchedTotal += 100.0f * (entry!!.watchTime / entry.duration)
```
Comment (ChrisSBras) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt:56]: Is the duration now used or is this still placeholder? I thought the duration was hard to read from a torrent, seeing that the function `updateTorrentDuration` is unused.

**Review by Floor97 — CHANGES_REQUESTED** [review]

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt (hunk 1 above)
Comment (junojense) [reply detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt:56]: Removed duration as we do not use it

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    val firstSeen: Long = 0
+)
+
+class Profile(
+    val magnets: HashMap<String, ProfileEntry> = HashMap()
+) {
+    fun updateEntryWatchTime(torrent: TorrentManager.TorrentHandler, time: Long) {
+        val name = torrent.torrentName + "[" + torrent.fileName + "]"
+        if(!magnets.contains(name)) magnets[name] =
+            ProfileEntry(0, time, System.currentTimeMillis())
```
[the remark this answers is not in the document]
Comment (junojense) [reply detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt:22]: This was the duration initially, but as I removed duration it should now be correct

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        if(!magnets.contains(name)) magnets[name] =
+            ProfileEntry(duration, 0, System.currentTimeMillis())
+        else magnets[name]!!.duration = duration
+        Log.i("DeToks", "Updated duration of $name to ${magnets[name]!!.duration}")
+    }
+}
+
+class Recommender {
+    private fun coinTossRecommender(URIs: List<String>): String {
+        return URIs[Random.nextInt(URIs.size)]
```
[the remark this answers is not in the document]
Comment (junojense) [reply detoks/src/main/java/nl/tudelft/trustchain/detoks/Recommender.kt:42]: Changed this to return a full hashmap which is just randomly shuffled

**Review by Floor97 — APPROVED** [review]
Lgtm!

**Review by ChrisSBras — APPROVED** [review]

**Pull request merged**